# A note about accessing the course files through Github

All of the lectures notes will be posted on the [class Github repo](https://github.com/iamwfx/4680_5680_intro_uds). 

We are (probably) not going over git and Github during class. If you're familiar with git/Github, feel free to clone the repo to get the new lecture materials for each class. Otherwise, I recommend you do the following: 
- Create class folder and name it `Intro_UDS`
- For each week, create a folder called `Week1`, `Week2`, etc. 
- To download the materials (Juputer notebook, data files, etc.), navigate to the file you want to download and select `Raw` above the view of the file. 
- Save this "raw" file in your class folder and **make sure it is in the proper file format**. For instance, if the file is `.ipynb` make sure you save the downloaded file as `.ipynb` (your computer might try to default to `.txt`.)

# Learning goals
After this week's lesson you should be able to:
- Explain what a Pandas Series is and how to select, filter, and replace values in the series 
- Read and explore tabular data in Python using a Pandas DataFrame
- Read and write datasets

This week's lessons are adapted from:
- [Geo-Python Lesson 5](https://geo-python-site.readthedocs.io/en/latest/lessons/L5/overview.html)
- [Practical Data Sciece on Pandas DataFrames](https://www.practicaldatascience.org/html/30_pandas_dataframes.html)

# Grading
As a reminder, your in-class exercises will be graded on completion.

# Installing Pandas
- First we'll need to install the `pandas` library if you are still using the `gds_py_reduced.yml` file that Jacob shared with you earlier today. 
- You can either do this in the Anaconda Navigator or 
- In your jupyter notebook, type 
```
    `!pip install pandas` 
```

In [6]:
# The ! in front of pip is a special Jupyter Notebook command that allows you to run terminal commands from the notebook.
!pip install pandas

In [7]:
### Let's first import the pandas library
### We will use the alias pd for pandas to make it easier to type. 
import pandas as pd 

# 0. What is Pandas? 

[Pandas](http://pandas.pydata.org/) is a widely used Python library for data analysis. 

### Easy-to-use data structures
In pandas, the data is typically stored in a data structure called a
DataFrame that looks like a typical table with rows and columns (+
indices and column names), where columns can contain data of different
data types. Thus, it is similar in some sense to how data is stored in
Excel or in R, which also uses a concept of a dataframe. In fact, Wes
McKinney first [developed pandas as an alternative for
R](https://blog.quantopian.com/meet-quantopians-newest-advisor-wes-mckinney/)
to deal with different complex data structures.

### Summary of the pandas data structure
The below is a picture of how data is structured in a dataframe. You can see it is tabular structure with rows, columns, and a (row) index, which is in many ways similar to what you might be familiar with from a Excel or Google Spreadsheets. Using this structure, we can use operations like arithmetic, columns and rows selection, columns and rows addition etc.

![Alt text](img/creating_dataframe1.png)

### Combines functionalities from many Python modules

pandas takes advantage of the [NumPy](http://www.numpy.org/) module
under the hood, which is mostly written in C. This makes it a fast and
powerful library that can efficiently handle even very large datasets.
pandas offers an easier and more intuitive syntax to do data analysis
and manipulation using either Numpy functionalities in the background or dedicated functionalities written explicitly for pandas.

 However, pandas is much more than an easier-to-use Numpy as it also combines many functionalities from other Python libraries such as [matplotlib (plotting)](https://matplotlib.org/) and [scipy(mathematics, science, engineering)](https://www.scipy.org/). Thus, you can use many of the features included in those packages without importing them at all.

### Supports data read/write from multiple formats
One of the most useful features of pandas is its ability to read data
from numerous different data formats directly. For example, pandas
supports reading and writing data from/to:

-   CSV
-   JSON
-   HTML
-   MS Excel
-   HDF5
-   Stata
-   SAS
-   Python Pickle format
-   SQL (Postgresql, MySQL, Oracle, MariaDB, etc.)

You can view the full list of supported data formats from the [pandas
docs](https://pandas.pydata.org/docs/user_guide/io.html).


# 1. DataFrames
Pandas DataFrames are tabular (as in table) data consisting of a collection of Series in which each column is a Pandas series. It is the central data structure used in most analysis using the Pandas library. 

First, let's read in our data using the pandas function `.read_csv()`. 

One of the nifty things about reading data in pandas is that it's designed to read many different types data sources, from files, to QL databases, and URLs. Here are a few to know about: 
- `pd.read_csv`: Read in a comma-separated-value file
- `pd.read_excel`: Read in an Excel (`.xls` and `.xlsx`) spreadsheet
- `pd.read_stata`: Read Stata (`.dta`) datasets
- `pd.read_hdf`: Read HDF (`.hdf`) datasets
- `pd.read_sql`: Read from a SQL database
- `pd.read_html`: Read from the `html` tags of an HTML file

Similarly, you can write a dataframe to many formats: (`df` here is the name of a dataframe)
- `df.to_csv`: Write to a comma-separated-value file
- `df.to_excel`: Write to an Excel (.xls and .xlsx) spreadsheet
- `df.to_stata`: Write to a stata (.dta) dataset
- `df.to_hdf`: Write to an HDF (.hdf) dataset
- `df.to_sql`: Write to a SQL database
- `df.to_html`: Write to an HTML table. 


Check out the [Pandas documentation on input/output](https://pandas.pydata.org/docs/reference/io.html) to see all the possible functions for reading data. 

## 1.1 Reading a file
Download the `msa_by_pop.csv` in this week's folder. We are first going to read this CSV as DataFrame into Pandas.

The function `.read_csv()` takes a file path as a string (read: text). If you have saved `msa_by_pop.csv` in the same folder as this notebook, then all you will need to input as your path is `msa_by_pop.csv`. 

(If you had saved your CSV within a sub-directory called `Data`, then to access this data file you'd need to input `Data/msa_by_pop.csv`)

In [8]:
msa_by_pop = pd.read_csv('msa_by_pop.csv')


In [9]:
msa_by_pop

Rank                                           MSA  population_2020
0       1  New York-Newark-Jersey City, NY-NJ-CT-PA MSA         20140470
1       2        Los Angeles-Long Beach-Anaheim, CA MSA         13200998
2       3        Chicago-Naperville-Elgin, IL-IN-WI MSA          9618502
3       4           Dallas-Fort Worth-Arlington, TX MSA          7637387
4       5      Houston-The Woodlands-Sugar Land, TX MSA          7122240
..    ...                                           ...              ...
379   380                              Danville, IL MSA            74188
380   381                           Lewiston, ID-WA MSA            64375
381   382                           Walla Walla, WA MSA            62584
382   383                                  Enid, OK MSA            62846
383   384                           Carson City, NV MSA            58639

[384 rows x 3 columns]

## 1.2 File Paths
`'msa_by_pop.csv'` is our file path. A file path is a string that tells the computer where to find a file. The file path above is a relative path, meaning it is relative to the location of this notebook. If you move this notebook to a different folder, the relative path will no longer work. An absolute path is a path that is not relative to the location of the notebook. An absolute path will work no matter where the notebook is located

In [10]:
# # For my computer, I can also read in the same dataset by using the absolute path.
# msa_by_pop = pd.read_csv('/Users/wenfeixu/Documents/GITHUB/4680_5680_intro_uds/Week1/msa_by_pop.csv')


In [11]:
msa_by_pop

Rank                                           MSA  population_2020
0       1  New York-Newark-Jersey City, NY-NJ-CT-PA MSA         20140470
1       2        Los Angeles-Long Beach-Anaheim, CA MSA         13200998
2       3        Chicago-Naperville-Elgin, IL-IN-WI MSA          9618502
3       4           Dallas-Fort Worth-Arlington, TX MSA          7637387
4       5      Houston-The Woodlands-Sugar Land, TX MSA          7122240
..    ...                                           ...              ...
379   380                              Danville, IL MSA            74188
380   381                           Lewiston, ID-WA MSA            64375
381   382                           Walla Walla, WA MSA            62584
382   383                                  Enid, OK MSA            62846
383   384                           Carson City, NV MSA            58639

[384 rows x 3 columns]

## 1.3 Exploring our data
You can see here that (in addition to the formatting of tabular data in Jupyter) the main difference between this and a series is that we have multiple columns with column labels. So the dataframe structure consists of: 
- An index, with index labels (here the labels are just `0`,`1`,...,`383`)
- Columns, with column labels (here `MSA`, `population_2020`,`perc_change`)
- And the data, which are the values in each row. 

We can use `.index` to get all the index values in the dataframe. 

In [12]:
msa_by_pop.index
## Ok, here it's just the RangeIndex, which is the default index for pandas dataframes. 
## Pandas isn't printing the entire index, but we can see that it starts at 0 and ends at 384.

RangeIndex(start=0, stop=384, step=1)

Now, in addition to `.index` we can also see all the columns in our DataFrame: 

In [13]:
msa_by_pop.columns

Index(['Rank', 'MSA', 'population_2020'], dtype='object')

Now, to see the datatypes for each column, we use `.dtypes`

In [14]:
msa_by_pop.dtypes

Rank                int64
MSA                object
population_2020     int64
dtype: object

One common function used to explore the data is called `.head()` that reveals the first 5 rows of the Dataframe. 

In [15]:
msa_by_pop.head()

Rank                                           MSA  population_2020
0     1  New York-Newark-Jersey City, NY-NJ-CT-PA MSA         20140470
1     2        Los Angeles-Long Beach-Anaheim, CA MSA         13200998
2     3        Chicago-Naperville-Elgin, IL-IN-WI MSA          9618502
3     4           Dallas-Fort Worth-Arlington, TX MSA          7637387
4     5      Houston-The Woodlands-Sugar Land, TX MSA          7122240

If you start to type `msa_by_pop.head(` without finishing the parens you can see the inputs required of the function: 


<img src="img/func_arg.png" alt="drawing" width="400" style="display: block; margin: 0 auto"/>


This shows us that that `.head()` by default show 5 rows, but you can also optionally adjust this by specifying another integer. For instance: 

In [16]:
## This gives us the first 10 rows
msa_by_pop.head(10)

Rank                                               MSA  population_2020
0     1      New York-Newark-Jersey City, NY-NJ-CT-PA MSA         20140470
1     2            Los Angeles-Long Beach-Anaheim, CA MSA         13200998
2     3            Chicago-Naperville-Elgin, IL-IN-WI MSA          9618502
3     4               Dallas-Fort Worth-Arlington, TX MSA          7637387
4     5          Houston-The Woodlands-Sugar Land, TX MSA          7122240
5     6  Washington-Arlington-Alexandria, DC-VA-MD-WV MSA          6385162
6     7   Philadelphia-Camden-Wilmington, PA-NJ-DE-MD MSA          6245051
7     8          Atlanta-Sandy Springs-Alpharetta, GA MSA          6089815
8     9       Miami-Fort Lauderdale-Pompano Beach, FL MSA          6138333
9    10                     Phoenix-Mesa-Chandler, AZ MSA          4845832

`.sample()` gives us a random selection of rows. 

In [17]:
msa_by_pop.sample(10)

Rank                              MSA  population_2020
137   138                   Peoria, IL MSA           402391
253   254              Bloomington, IL MSA           170954
123   124                  Salinas, CA MSA           439035
270   271  Elizabethtown-Fort Knox, KY MSA           155572
88     89                  Madison, WI MSA           680796
176   177                     Waco, TX MSA           277547
30     31           Kansas City, MO-KS MSA          2192035
93     94                   Toledo, OH MSA           646604
234   235                 Kingston, NY MSA           181851
223   224                     Bend, OR MSA           198253

The function `.shape` gives us the number of `(rows,columns)` in our dataframe: 

In [18]:
msa_by_pop.shape

(384, 3)

The function `.len()` measures the length of a selection. Applying this to the DataFrame gives you the number of rows: 

In [19]:
len(msa_by_pop)

384

Applying it to the columns gives us the number of columns we have

In [20]:
len(msa_by_pop.columns)

3

`.describe()` provides some basic descriptive statistics for our dataframe: 

In [21]:
msa_by_pop.describe()

Rank  population_2020
count  384.000000     3.840000e+02
mean   192.500000     7.450639e+05
std    110.995495     1.670046e+06
min      1.000000     5.863900e+04
25%     96.750000     1.474762e+05
50%    192.500000     2.484905e+05
75%    288.250000     5.956365e+05
max    384.000000     2.014047e+07

`.sort-values()` sorts your DataFrame by a certain column. If you column is numeric, it will sort the values from smallest to largest. If your column is a string, it will sort alphabetically.

The index values will be sorted along with the rows. 

In [22]:
# This sorts the dataframe by the population_2021_est column, from smallest to largest
# Note that the original dataframe is not changed 
# but that the index values now reflect the sorted order. 
msa_by_pop.sort_values('population_2020')

Rank                                           MSA  population_2020
383   384                           Carson City, NV MSA            58639
381   382                           Walla Walla, WA MSA            62584
382   383                                  Enid, OK MSA            62846
380   381                           Lewiston, ID-WA MSA            64375
379   380                              Danville, IL MSA            74188
..    ...                                           ...              ...
4       5      Houston-The Woodlands-Sugar Land, TX MSA          7122240
3       4           Dallas-Fort Worth-Arlington, TX MSA          7637387
2       3        Chicago-Naperville-Elgin, IL-IN-WI MSA          9618502
1       2        Los Angeles-Long Beach-Anaheim, CA MSA         13200998
0       1  New York-Newark-Jersey City, NY-NJ-CT-PA MSA         20140470

[384 rows x 3 columns]

Adding the `ascending=False` input in your function will sort from largest to smallest or end of alphabet to beginning, depending on your column data type. 

In [23]:
msa_by_pop.sort_values('population_2020',ascending=False)

Rank                                           MSA  population_2020
0       1  New York-Newark-Jersey City, NY-NJ-CT-PA MSA         20140470
1       2        Los Angeles-Long Beach-Anaheim, CA MSA         13200998
2       3        Chicago-Naperville-Elgin, IL-IN-WI MSA          9618502
3       4           Dallas-Fort Worth-Arlington, TX MSA          7637387
4       5      Houston-The Woodlands-Sugar Land, TX MSA          7122240
..    ...                                           ...              ...
379   380                              Danville, IL MSA            74188
380   381                           Lewiston, ID-WA MSA            64375
382   383                                  Enid, OK MSA            62846
381   382                           Walla Walla, WA MSA            62584
383   384                           Carson City, NV MSA            58639

[384 rows x 3 columns]

## 2.4 Selecting columns and rows
Since we have column and row indexes, we can select subsets of the dataframe using the .loc method.
Let's select the row with index value 0, and all columns

In [24]:
msa_by_pop.loc[0,:]

Rank                                                          1
MSA                New York-Newark-Jersey City, NY-NJ-CT-PA MSA
population_2020                                        20140470
Name: 0, dtype: object

For `.iloc` we can select rows and columns now by their position.

In [25]:
# Remember integer ranges are exclusive of the last value!
msa_by_pop.iloc[0:5,0:3]

Rank                                           MSA  population_2020
0     1  New York-Newark-Jersey City, NY-NJ-CT-PA MSA         20140470
1     2        Los Angeles-Long Beach-Anaheim, CA MSA         13200998
2     3        Chicago-Naperville-Elgin, IL-IN-WI MSA          9618502
3     4           Dallas-Fort Worth-Arlington, TX MSA          7637387
4     5      Houston-The Woodlands-Sugar Land, TX MSA          7122240

Similarly, using `.loc` we can select rows and columns by their names. 

In [26]:
# Since our rows are index by integers, when we can use the .loc method to select rows, 
# it's the same as .iloc
msa_by_pop.loc[0:5,['Rank','MSA','population_2020']]

Rank                                               MSA  population_2020
0     1      New York-Newark-Jersey City, NY-NJ-CT-PA MSA         20140470
1     2            Los Angeles-Long Beach-Anaheim, CA MSA         13200998
2     3            Chicago-Naperville-Elgin, IL-IN-WI MSA          9618502
3     4               Dallas-Fort Worth-Arlington, TX MSA          7637387
4     5          Houston-The Woodlands-Sugar Land, TX MSA          7122240
5     6  Washington-Arlington-Alexandria, DC-VA-MD-WV MSA          6385162

You can always leave out the column argument in either `.loc` or `.iloc` in order to select all columns. 

In [27]:
msa_by_pop.iloc[0:5]

Rank                                           MSA  population_2020
0     1  New York-Newark-Jersey City, NY-NJ-CT-PA MSA         20140470
1     2        Los Angeles-Long Beach-Anaheim, CA MSA         13200998
2     3        Chicago-Naperville-Elgin, IL-IN-WI MSA          9618502
3     4           Dallas-Fort Worth-Arlington, TX MSA          7637387
4     5      Houston-The Woodlands-Sugar Land, TX MSA          7122240

In [28]:
msa_by_pop.loc[0:5]

Rank                                               MSA  population_2020
0     1      New York-Newark-Jersey City, NY-NJ-CT-PA MSA         20140470
1     2            Los Angeles-Long Beach-Anaheim, CA MSA         13200998
2     3            Chicago-Naperville-Elgin, IL-IN-WI MSA          9618502
3     4               Dallas-Fort Worth-Arlington, TX MSA          7637387
4     5          Houston-The Woodlands-Sugar Land, TX MSA          7122240
5     6  Washington-Arlington-Alexandria, DC-VA-MD-WV MSA          6385162

But to do the same with columns you'll first have to specify that you want all the rows with `:` . 

In [29]:
msa_by_pop.loc[:,['Rank']]

Rank
0       1
1       2
2       3
3       4
4       5
..    ...
379   380
380   381
381   382
382   383
383   384

[384 rows x 1 columns]

In [30]:
msa_by_pop.iloc[:,[0]]

Rank
0       1
1       2
2       3
3       4
4       5
..    ...
379   380
380   381
381   382
382   383
383   384

[384 rows x 1 columns]

With the square brackets `[]` you can provide a row or column. 

In [31]:
msa_by_pop['Rank']

0        1
1        2
2        3
3        4
4        5
      ... 
379    380
380    381
381    382
382    383
383    384
Name: Rank, Length: 384, dtype: int64

If you put the column name in a list, you get back a dataframe instead of a series. 

In [32]:
msa_by_pop[['Rank']]

Rank
0       1
1       2
2       3
3       4
4       5
..    ...
379   380
380   381
381   382
382   383
383   384

[384 rows x 1 columns]

This list can have more than one element, of course. 

In [33]:
msa_by_pop[['Rank','population_2020']]

Rank  population_2020
0       1         20140470
1       2         13200998
2       3          9618502
3       4          7637387
4       5          7122240
..    ...              ...
379   380            74188
380   381            64375
381   382            62584
382   383            62846
383   384            58639

[384 rows x 2 columns]

## 2.5 Filtering
We can filter the dataframe by expressions. Filtering in pandas is done with a boolean expression. The expression is evaluated for each row in the dataframe, and only rows where the expression evaluates to `True` are returned.

In [34]:
# A boolean expression is one that evaluates to either True or False.
# Here we are asking if the value in the Rank column is greater than 300.

msa_by_pop.loc[msa_by_pop['Rank']>300]
# Yes, msa_by_pop['Rank'] does also select the column, 
# but notice that it gives a Series instead of a dataframe one column.

Rank                   MSA  population_2020
300   301          Napa, CA MSA           138019
301   302  Harrisonburg, VA MSA           135571
302   303        Sumter, SC MSA           136700
303   304  The Villages, FL MSA           129752
304   305   Springfield, OH MSA           136001
..    ...                   ...              ...
379   380      Danville, IL MSA            74188
380   381   Lewiston, ID-WA MSA            64375
381   382   Walla Walla, WA MSA            62584
382   383          Enid, OK MSA            62846
383   384   Carson City, NV MSA            58639

[84 rows x 3 columns]

In [35]:
# This expression finds all the rows where the value in the Rank column is greater than 300,
msa_by_pop['Rank']>300
# 

0      False
1      False
2      False
3      False
4      False
       ...  
379     True
380     True
381     True
382     True
383     True
Name: Rank, Length: 384, dtype: bool

In order to filter by more than one condition, you must: 
1. Put all conditions in `()`
2. Separate the condtions by: 

    a. `|` if an `OR` condition     
    b. `&` if an `AND` condition

In [36]:
msa_by_pop[(msa_by_pop['Rank']>300) & (msa_by_pop['MSA'].str.contains('NY'))]

Rank                          MSA  population_2020
317   318          Glens Falls, NY MSA           127039
334   335  Watertown-Fort Drum, NY MSA           116721
346   347               Ithaca, NY MSA           105740
373   374               Elmira, NY MSA            84148

In [37]:
msa_by_pop[(msa_by_pop['Rank']>300) | (msa_by_pop['MSA'].str.contains('NY'))]

Rank                                           MSA  population_2020
0       1  New York-Newark-Jersey City, NY-NJ-CT-PA MSA         20140470
48     49                 Buffalo-Niagara Falls, NY MSA          1166902
51     52                             Rochester, NY MSA          1090135
62     63               Albany-Schenectady-Troy, NY MSA           899262
83     84      Poughkeepsie-Newburgh-Middletown, NY MSA           697221
..    ...                                           ...              ...
379   380                              Danville, IL MSA            74188
380   381                           Lewiston, ID-WA MSA            64375
381   382                           Walla Walla, WA MSA            62584
382   383                                  Enid, OK MSA            62846
383   384                           Carson City, NV MSA            58639

[93 rows x 3 columns]

## 2.6 Modifying elements
Essentially, in the same way that we can select elements we can also update them using the same logic. 

In [38]:
### Filter for population value
msa_by_pop.loc[msa_by_pop['population_2020']==20140470,'population_2020']

0    20140470
Name: population_2020, dtype: int64

In [39]:
# This is the same as above, but we are assigning the value 20000000 to the population_2021_est column
msa_by_pop.loc[msa_by_pop['population_2020']==20140470,'population_2020'] = 20000000

## 2.7 Pandas Arithmetics
Pandas arithmetics are vectorized, meaning that the operation is applied to each element in the column.

In [40]:
# Pandas arithmetics are vectorized, meaning that the operation is applied to each element in the column
msa_by_pop['population_2020'] / 1000000

0      20.000000
1      13.200998
2       9.618502
3       7.637387
4       7.122240
         ...    
379     0.074188
380     0.064375
381     0.062584
382     0.062846
383     0.058639
Name: population_2020, Length: 384, dtype: float64

In [41]:
# We can assign the result of this operation to a new column
msa_by_pop['population_2020_millions'] = msa_by_pop['population_2020'] / 1000000

In [42]:
# This multiplies the population_2020 column by 2 
msa_by_pop['population_2020'] * 2

0      40000000
1      26401996
2      19237004
3      15274774
4      14244480
         ...   
379      148376
380      128750
381      125168
382      125692
383      117278
Name: population_2020, Length: 384, dtype: int64

In [43]:
# This takes the square of the population_2020 column
msa_by_pop['population_2020'] **2

0      400000000000000
1      174266348196004
2       92515580724004
3       58329680187769
4       50726302617600
            ...       
379         5503859344
380         4144140625
381         3916757056
382         3949619716
383         3438532321
Name: population_2020, Length: 384, dtype: int64

We can find the sum by using the sum method. Note that the sum method ignores missing values by default

In [44]:
msa_by_pop['population_2020'].sum()

285964086

In [45]:
msa_by_pop['population_2020'].mean()

744698.140625

In [46]:
msa_by_pop['population_2020'].median()

248490.5

In [47]:
msa_by_pop['population_2020'].std()

1665797.0023611302

# 3. Good Coding Practices

The following are some good practices for writing more legible Jupyter notebooks. Often, we don't necessarily realize that code we write isn't immediately interpretable to readers. To make code more easily interpreted, we will often explain through markdown text or comments what we are doing. 

## 3.1 Markdown and explanatory cells
As you can see in this notebook, there are many "Markdown" cells surrounding our actual code. The markdown I have here describes the purpose of each code cell and what I wanted to do with it. 

I often organize my notebooks by header size `#`, `##` etc, and by numbering the different sections. This can be helpful if you're writing an especially long notebook. 

Here's a [guide on how to write in Markdown](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax).

## 3.2 Indents
Indents are very important in python. 


In [49]:
# Indents are important in Python! 
# They tell Python what code is part of a function or loop.
# This is a for loop.
for i in range(5):
    print(i)
# This will give you an error: 
for i in range(5):
    print(i)


0
1
2
3
4
0
1
2
3
4


## 3.3 Commenting
Code comments are lines that start with a #. They are not executed by Python. They are used to explain what the code does.

You can see that sometimes I have small notes in there that maybe didn't need to go into the markdown cell. 

In [ ]:
# This is a code comment
msa_by_pop[(msa_by_pop['Rank']>300) & (msa_by_pop['MSA'].str.contains('NY')) & (msa_by_pop['population_2020']>100000)]

## 3.4 Selecting variable names
There are two aspects of selecting variable names we're going to go over today: 
1. What to name your variables or function
2. How to name a multi-word variable or function

### 3.3.1
A good variable name should: 
- Be clear and concise.
- Be written in English. A general coding practice is to write code with variable names in English, as that is the most likely common language between programmers. Thus, variable names such as muuttuja (which is also not a good name on other levels) should be avoided.
- Not contain special characters. Python supports use of special characters by way of various encoding options that can be given in a program. That said, it is better to avoid variables such as lämpötila because encoding issues can arise in some cases. Better to stick to the standard printable ASCII character set to be safe.
- Not conflict with any Python keywords, such as for, True, False, and, if, or else. These are reserved for special operations in Python and cannot be used as variable names.

In [ ]:
# Do not do this: 
finnishmeteorlogicalinstituteobservationstationidentificationnumber = "101533"


In [ ]:
# Or this: 
f = "101533"

In [ ]:
# Something that is as short as possible while still being descriptive is best
sid = "101533"

### 3.3.2 Snake case and camel case
There are two general ways of connecting variable and function names that contain more than one word. 

You can see from the above exercise that I've named my DataFrame `msa_by_pop`. Connecting words by `_`  is called "Snake Case". 

Another convention is to use capital letters to start new words. `MsaByPop` could be another way to name our variable. 

# 4. In-Class Exercises
Each week, we will have in-class exercises that give you some pratice on the concepts learned in class.


## 4.1. Exercise 1 
We are going to use the `msa_by_pop` data frame we created above. 

In [51]:
import pandas as pd
msa_by_pop = pd.read_html("https://en.wikipedia.org/wiki/Metropolitan_statistical_area")[1]
msa_by_pop = msa_by_pop.reset_index().iloc[1:, :]
msa_by_pop.columns =['Rank', 'MSA', 'population_2021_est', 'population_2020',
       'perc_change', 'Encompassing_combined_statistical_area']

Rank                                     MSA  population_2021_est  \
1       1  New York–Newark–Jersey City, NY-NJ MSA           19557311.0   
2       2  Los Angeles–Long Beach–Anaheim, CA MSA           12872322.0   
3       3     Chicago–Naperville–Elgin, IL-IN MSA            9274140.0   
4       4     Dallas–Fort Worth–Arlington, TX MSA            7943685.0   
5       5  Houston–Pasadena–The Woodlands, TX MSA            7368466.0   
..    ...                                     ...                  ...   
383   383                     Lewiston, ID-WA MSA              65512.0   
384   384                            Enid, OK MSA              61920.0   
385   385                     Walla Walla, WA MSA              61890.0   
386   386                     Carson City, NV MSA              58130.0   
387   387                          Eagle Pass, TX              57843.0   

     population_2020 perc_change  \
1         20081935.0      −2.61%   
2         13200998.0      −2.49%   
3          9449351.0      −1.85%   
4          7637387.0      +4.01%   
5          7149642.0      +3.06%   
..               ...         ...   
383          64375.0      +1.77%   
384          62846.0      −1.47%   
385          62584.0      −1.11%   
386          58639.0      −0.87%   
387          57887.0      −0.08%   

               Encompassing_combined_statistical_area  
1                    New York–Newark, NY-NJ-CT-PA CSA  
2                      Los Angeles–Long Beach, CA CSA  
3                    Chicago–Naperville, IL-IN-WI CSA  
4                        Dallas–Fort Worth, TX-OK CSA  
5                            Houston–Pasadena, TX CSA  
..                                                ...  
383                                               NaN  
384                                               NaN  
385            Kennewick–Richland-Walla Walla, WA CSA  
386  Reno–Carson City–Gardnerville Ranchos, NV-CA CSA  
387                                               NaN  

[387 rows x 6 columns]

Which is the MSA with the 10th largest population for 2020? 

In [71]:
## Insert your code here
msa_by_pop_2021_descend=msa_by_pop.sort_values('population_2021_est',ascending=False)
print("2021 10th largest population:", msa_by_pop_2021_descend.loc[10, "MSA"])

msa_by_pop_2020_descend=msa_by_pop.sort_values('population_2020',ascending=False)
print("2020 10th largest population:", msa_by_pop_2020_descend.loc[10, "MSA"])

2021 10th largest population: Phoenix–Mesa–Chandler, AZ MSA
2020 10th largest population: Phoenix–Mesa–Chandler, AZ MSA


Select the smallest 10 MSAs by 2020 Census population

In [81]:
## Insert your code here
msa_by_pop_2020_ascend=msa_by_pop.sort_values('population_2020')
msa_by_pop_2020_ascend.iloc[0:10]

Rank                   MSA  population_2021_est  population_2020  \
387   387        Eagle Pass, TX              57843.0          57887.0   
386   386   Carson City, NV MSA              58130.0          58639.0   
385   385   Walla Walla, WA MSA              61890.0          62584.0   
384   384          Enid, OK MSA              61920.0          62846.0   
383   383   Lewiston, ID-WA MSA              65512.0          64375.0   
381   381  Grand Island, NE MSA              76333.0          77038.0   
382   382         Minot, ND MSA              76279.0          77546.0   
380   380        Casper, WY MSA              79601.0          79955.0   
374   374    Hinesville, GA MSA              86378.0          81424.0   
378   378      Columbus, IN MSA              83540.0          82208.0   

    perc_change            Encompassing_combined_statistical_area  
387      −0.08%                                               NaN  
386      −0.87%  Reno–Carson City–Gardnerville Ranchos, NV-CA CSA  
385      −1.11%            Kennewick–Richland-Walla Walla, WA CSA  
384      −1.47%                                               NaN  
383      +1.77%                                               NaN  
381      −0.92%                                               NaN  
382      −1.63%                                               NaN  
380      −0.44%                                               NaN  
374      +6.08%            Savannah–Hinesville–Statesboro, GA CSA  
378      +1.62%                Indianapolis–Carmel–Muncie, IN CSA

Find the total 2020 Census population of these 384 MSAs. 

In [83]:
## Insert your code here
msa_by_pop['population_2020'].sum()

285649602.0

Based on the population in 2020 and assuming a 2% population increase across all cities, let's estimate the population in 2022 using the following function: 

$$
pop_{year2020+t} =pop_{year2020}*(1+ {\% change})^t
$$


where  $\% change$ is positive 2%. 


First, what's t? 

In [ ]:
## Insert your code here
t= 2

Now create another column called `population_2022_estimate` and estimate the 2022 population counts using the formula above. 

In [84]:
## Insert your code here
msa_by_pop['population_2022_estimate']= msa_by_pop['population_2020']*(1+0.02)**2

In [85]:
msa_by_pop

Rank                                     MSA  population_2021_est  \
1       1  New York–Newark–Jersey City, NY-NJ MSA           19557311.0   
2       2  Los Angeles–Long Beach–Anaheim, CA MSA           12872322.0   
3       3     Chicago–Naperville–Elgin, IL-IN MSA            9274140.0   
4       4     Dallas–Fort Worth–Arlington, TX MSA            7943685.0   
5       5  Houston–Pasadena–The Woodlands, TX MSA            7368466.0   
..    ...                                     ...                  ...   
383   383                     Lewiston, ID-WA MSA              65512.0   
384   384                            Enid, OK MSA              61920.0   
385   385                     Walla Walla, WA MSA              61890.0   
386   386                     Carson City, NV MSA              58130.0   
387   387                          Eagle Pass, TX              57843.0   

     population_2020 perc_change  \
1         20081935.0      −2.61%   
2         13200998.0      −2.49%   
3          9449351.0      −1.85%   
4          7637387.0      +4.01%   
5          7149642.0      +3.06%   
..               ...         ...   
383          64375.0      +1.77%   
384          62846.0      −1.47%   
385          62584.0      −1.11%   
386          58639.0      −0.87%   
387          57887.0      −0.08%   

               Encompassing_combined_statistical_area  \
1                    New York–Newark, NY-NJ-CT-PA CSA   
2                      Los Angeles–Long Beach, CA CSA   
3                    Chicago–Naperville, IL-IN-WI CSA   
4                        Dallas–Fort Worth, TX-OK CSA   
5                            Houston–Pasadena, TX CSA   
..                                                ...   
383                                               NaN   
384                                               NaN   
385            Kennewick–Richland-Walla Walla, WA CSA   
386  Reno–Carson City–Gardnerville Ranchos, NV-CA CSA   
387                                               NaN   

     population_2022_estimate  
1                2.089325e+07  
2                1.373432e+07  
3                9.831105e+06  
4                7.945937e+06  
5                7.438488e+06  
..                        ...  
383              6.697575e+04  
384              6.538498e+04  
385              6.511239e+04  
386              6.100802e+04  
387              6.022563e+04  

[387 rows x 7 columns]